sstables/sstables_manager.cc

```diff
@@ -74,16 +75,26 @@ future<> storage_manager::stop() {
 }
 
 future<> storage_manager::update_config(const db::config& cfg) {
-    for (auto [ep, ecfg] : cfg.object_storage_config()) {
-        auto s3_cfg = make_lw_shared<s3::endpoint_config>(std::move(ecfg));
-        auto [it, added] = _s3_endpoints.try_emplace(ep, std::move(s3_cfg));
-        if (!added) {
+    // Updates S3 client configurations if the endpoint is already known and
+    // removes the entries that are not present in the new configuration.
+    // Even though we remove obsolete S3 clients from this map, each IO
+    // holds a shared_ptr to the client, so the clients will be kept alive for
+    // as long as needed. 
+    std::unordered_map<sstring, s3_endpoint> updates;
+    for (auto& e : cfg.object_storage_endpoints()) {
+        auto s3_cfg = make_lw_shared<s3::endpoint_config>(std::move(e.config));
+        auto existing = _s3_endpoints.find(e.endpoint);
+        if (existing != _s3_endpoints.end()) {
+            auto [it, _] = updates.emplace(e.endpoint, existing->second);
             if (it->second.client != nullptr) {
```

Comment:
Is this `if` ever resolves to true? The it points to freshly implaced entry into the `updates` map, which starts from empty and `storage_manager::get_endpoint_client()` is not yet called on it

Reply:
the client is on `s3_endpoint` struct, so even though we emplace into updates, the `s3_endpoint::client` will point to the same instance that `_s3_endpoints` had. So the `if` I inherited from the original code should model the same case IIUC.

Reply:
Ah, you're right. The `auto [it, _] = updates.emplace(e.endpoint, existing->second);` above makes a copy of full `s3_endpoint` thus carrying the client pointer copy to `updates`.